tonic-build/src/manual.rs

```diff
@@ -0,0 +1,444 @@
+    server_streaming: bool,
+    /// The path to the codec to use for this method
+    codec_path: Option<String>,
+}
+
+impl MethodBuilder {
+    /// Set the name for this Method.
+    ///
+    /// This value will be used for generating the client functions for calling this Method.
+    ///
+    /// Generally this is formatted in snake_case.
+    pub fn name<T: AsRef<str>>(mut self, name: T) -> Self {
+        self.name = Some(name.as_ref().to_owned());
+        self
+    }
+
+    /// Set the route_name for this Method.
+    ///
+    /// This value will be used as part of the route for calling this method.
+    /// Routes have the form: `/<package_name>.<service_name>/<method_route_name>`
+    ///
+    /// Generally this is formatted in PascalCase.
+    pub fn route_name<T: AsRef<str>>(mut self, route_name: T) -> Self {
+        self.route_name = Some(route_name.as_ref().to_owned());
+        self
+    }
+
+    /// Add a comment string that should be included as a doc comment for this Method.
+    pub fn comment<T: AsRef<str>>(mut self, comment: T) -> Self {
+        self.comments.push(comment.as_ref().to_owned());
+        self
+    }
+
+    /// Set the path to the Rust type that should be use for the Request type of this method.
+    pub fn input_type<T: AsRef<str>>(mut self, input_type: T) -> Self {
+        self.input_type = Some(input_type.as_ref().to_owned());
+        self
+    }
+
+    /// Set the path to the Rust type that should be use for the Response type of this method.
+    pub fn output_type<T: AsRef<str>>(mut self, output_type: T) -> Self {
+        self.output_type = Some(output_type.as_ref().to_owned());
+        self
+    }
+
+    /// Set the path to the Rust type that should be used as the `Codec` for this method.
+    ///
+    /// Currently the codegen assumes that this type implements `Default`.
+    pub fn codec_path<T: AsRef<str>>(mut self, codec_path: T) -> Self {
+        self.codec_path = Some(codec_path.as_ref().to_owned());
+        self
+    }
+
+    /// Sets if the Method request from the client is streamed.
+    pub fn client_streaming(mut self) -> Self {
+        self.client_streaming = true;
+        self
+    }
+
+    /// Sets if the Method response from the server is streamed.
+    pub fn server_streaming(mut self) -> Self {
+        self.server_streaming = true;
+        self
+    }
+
+    /// Build a Method
+    ///
+    /// Panics if `name`, `route_name`, `input_type`, `output_type`, or `codec_path` weren't set.
+    pub fn build(self) -> Method {
+        Method {
+            name: self.name.unwrap(),
+            route_name: self.route_name.unwrap(),
+            comments: self.comments,
+            input_type: self.input_type.unwrap(),
+            output_type: self.output_type.unwrap(),
+            client_streaming: self.client_streaming,
+            server_streaming: self.server_streaming,
+            codec_path: self.codec_path.unwrap(),
+        }
+    }
+}
+
+struct ServiceGenerator {
```

Comment:
Do we actually need this extra struct? I don't mind keeping it just seems like we could just embed the code in the `compile` method?